File: user_player.rb
require_relative 'player'

module TicTacToe
  class UserPlayer < Player
    
    def select_position!
      @game.print_board
      loop do
        print "Select your #{marker} position"
        selection = gets.to_i
        return selection if @game.free_positions.include?(selection)
        puts "Position #{selection} is not available. Try again:"
      end
    end
    
    def to_s
      "User"
    end
  end
end

File: JavaTicTacToe/UserPlayer.java
import java.util.InputMismatchException;
import java.util.Scanner;

public class UserPlayer extends Player{
    private Scanner scanner;

    public UserPlayer(game game , String marker) {
        super(game, marker);
        this.scanner = new Scanner(System.in);
    }

    @Override 
    public int selectPosition() {
        while(true) {
            System.out.println("Select " + marker + " position:- ");
            try {
                int choice = scanner.nextInt();
                if ((game.getFreePositions().contains(choice))) {
                    return choice;
                }
                System.out.println("Position " + choice + " is not available. Try again.");
            } catch (InputMismatchException e) {
                System.out.println("Enter valid position");
                scanner.next();
            }  
        }  
    }

    @Override
    public String toString() {
        return "Human";
    }
}
